autogpt_platform/backend/schema.prisma

```diff
@@ -110,6 +112,57 @@ model AgentPreset {
   @@index([userId])
 }
 
+enum NotificationType {
+  AGENT_RUN
+  ZERO_BALANCE
+  LOW_BALANCE
+  BLOCK_EXECUTION_FAILED
+  CONTINUOUS_AGENT_ERROR
+  DAILY_SUMMARY
+  WEEKLY_SUMMARY
+  MONTHLY_SUMMARY
+}
+
+model NotificationEvent {
+  id        String   @id @default(uuid())
+  createdAt DateTime @default(now())
+  updatedAt DateTime @default(now()) @updatedAt
+
+  UserNotificationBatch   UserNotificationBatch? @relation(fields: [userNotificationBatchId], references: [id])
+  userNotificationBatchId String?
+
+  type NotificationType
+  data Json
+}
+
+model UserNotificationBatch {
+  id        String   @id @default(uuid())
+  createdAt DateTime @default(now())
+  updatedAt DateTime @default(now()) @updatedAt
+
+  userId String
+  user   User   @relation(fields: [userId], references: [id], onDelete: Cascade)
+
+  type NotificationType
+
+  notifications NotificationEvent[]
+
+  // Each user can only have one batch of a notification type at a time
+  @@unique([userId, type])
+}
+
+model UserNotificationPreference {
+  id        String   @id @default(uuid())
+  createdAt DateTime @default(now())
+  updatedAt DateTime @default(now()) @updatedAt
+
+  userId String
+  user   User   @relation(fields: [userId], references: [id], onDelete: Cascade)
+
+  // Dict[NotificationType, bool]
+  preferences Json @default("{}")
```

Comment:
Using json here seems unnecessary. Doing something like this seems more efficient for a database
```
  notify_on_AGENT_RUN
  notify_on_ZERO_BALANCE
  notify_on_LOW_BALANCE
  notify_on_BLOCK_EXECUTION_FAILED
  notify_on_CONTINUOUS_AGENT_ERROR
  notify_on_DAILY_SUMMARY
  notify_on_WEEKLY_SUMMARY
  notify_on_MONTHLY_SUMMARY
  ```

Reply:
Happy to make this change. Are we okay with migrations being needed for every new email type?

Reply:
Also should we auto sign everyone up for emails or no

Reply:
going with yes

Reply:
Yeah I think required a migration whenever we add a new notification type is acceptable

Reply:
okay cool, will make the change and include the pr smash